keychain/btcwallet.go

```diff
@@ -0,0 +1,283 @@
+// our keys. If the scope has already been fetched from the database, then a
+// ached version will be returned. Otherwise, we'll fetch it from the database
+// and cache it for subsequent accesses.
+func (b *BtcWalletKeyRing) keyScope() (*waddrmgr.ScopedKeyManager, error) {
+	// If the scope has already been populated, then we'll return it
+	// directly.
+	if b.lightningScope != nil {
+		return b.lightningScope, nil
+	}
+
+	// Otherwise, we'll first do a check to ensure that the root manager
+	// isn't locked, as otherwise we won't be able to *use* the scope.
+	if b.wallet.Manager.Locked() {
+		return nil, fmt.Errorf("cannot create BtcWalletKeyRing with " +
+			"locked waddrmgr.Manager")
+	}
+
+	// If the manager is indeed unlocked, then we'll fetch the scope, cache
+	// it, and return to the caller.
+	lnScope, err := b.wallet.Manager.FetchScopedKeyManager(
+		lightningKeyScope,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	b.lightningScope = lnScope
+
+	return lnScope, nil
+}
+
+// createAccountIfNotExists will create the corresponding account for a key
+// family if it doesn't already exist in the database.
+func (b *BtcWalletKeyRing) createAccountIfNotExists(
+	addrmgrNs walletdb.ReadWriteBucket, keyFam KeyFamily,
+	scope *waddrmgr.ScopedKeyManager) error {
+
+	// If this is the multi-sig key family, then we can return early as
+	// this is the default account that's created.
+	if keyFam == KeyFamilyMultiSig {
+		return nil
+	}
+
+	// Otherwise, we'll check if the account already exists, if so, we can
+	// once again bail early.
+	_, err := scope.AccountName(addrmgrNs, uint32(keyFam))
+	if err == nil {
+		return nil
+	}
+
+	// If we reach this point, then the account hasn't yet been created, so
+	// we'll need to create it before we can proceed.
+	return scope.NewRawAccount(addrmgrNs, uint32(keyFam))
+}
+
+// DeriveNextKey attempts to derive the *next* key within the key family
+// (account in BIP43) specified. This method should return the next external
+// child within this branch.
+//
+// NOTE: This is part of the keychain.KeyRing interface.
+func (b *BtcWalletKeyRing) DeriveNextKey(keyFam KeyFamily) (KeyDescriptor, error) {
+	var pubKey *btcec.PublicKey
+
+	db := b.wallet.Database()
+	err := walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+		addrmgrNs := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+
+		scope, err := b.keyScope()
+		if err != nil {
+			return err
+		}
+
+		// If the account doesn't exist, then we may need to create it
+		// for the first time in order to derive the keys that we
+		// require.
+		err = b.createAccountIfNotExists(addrmgrNs, keyFam, scope)
+		if err != nil {
+			return err
+		}
+
+		addrs, err := scope.NextExternalAddresses(
+			addrmgrNs, uint32(keyFam), 1,
+		)
+		if err != nil {
+			return err
+		}
+
+		pubKey = addrs[0].(waddrmgr.ManagedPubKeyAddress).PubKey()
+
+		return nil
+	})
+	if err != nil {
+		return KeyDescriptor{}, err
+	}
+
+	return KeyDescriptor{
```

Comment:
Why don't we include a locator for keys generated by this method?

Reply:
Fixed.

Reply:
Atm, we don't have a way of knowing _which_ child key was derived. This can be fixed in the future with a change to `btcwallet` itself though.